### PR TITLE
[InstCombine] Generalize and consolidate phi translation check

### PR DESCRIPTION
**Pull request by nikic**
The foldOpIntoPhi() transforms requires all operands to be phi-translatable. This can be the case either because they are phi nodes in the same block, or because the operand dominates the block.

Currently, most callers of foldOpIntoPhi() satisfy this pre-condition by requiring a constant operand, which trivially dominates everything. Only selects had handling for variable operands.

Move this logic into foldOpIntoPhi(), so things are handled correctly if other callers are generalized. Also make the implementation a bit more general by querying the dominator tree.

No compile-time impact but improvements on stage2 builds, so this helps clang: https://llvm-compile-time-tracker.com/compare.php?from=32679e10a9b66405c340213993f65b2edf5a794a&to=e99c004d372891fa23b79de92003fe123143dac2&stat=instructions:u

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 1]
```diff
@@ -1785,9 +1784,31 @@ Instruction *InstCombinerImpl::foldOpIntoPhi(Instruction &I, PHINode *PN) {
...
   }
 
+  // Check that all operands are phi-translatable.
+  for (Value *Op : I.operands()) {
+    if (Op == PN)
+      continue;
+
+    // Non-instructions never require phi-translation.
+    auto *I = dyn_cast<Instruction>(Op);
+    if (!I)
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1794]: Can you give this a different name (maybe `OpI`), the shadow variable is very confusing.

**Comment by nikic** [comment]
ping :)

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 2]
```diff
@@ -1785,9 +1784,31 @@ Instruction *InstCombinerImpl::foldOpIntoPhi(Instruction &I, PHINode *PN) {
...
+  for (Value *Op : I.operands()) {
+    if (Op == PN)
+      continue;
+
+    // Non-instructions never require phi-translation.
+    auto *OpI = dyn_cast<Instruction>(Op);
+    if (!OpI)
+      continue;
+
+    // Can phi-translate phi nodes in the same block.
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1797]: nit: "Can phi-translate..." -> "Phi-translate can handle..."

**Review by goldsteinn — APPROVED** [review]
LGTM

**Pull request merged**